contracts/FeeDisbursal/FeeDisbursal.sol

```diff
@@ -0,0 +1,269 @@
+pragma solidity 0.4.25;
+
+import "../Universe.sol";
+import "../AccessControl/AccessControlled.sol";
+import "../Reclaimable.sol";
+import "../Standards/IFeeDisbursal.sol";
+import "../Identity/IIdentityRegistry.sol";
+import "../SnapshotToken/BasicSnapshotToken.sol";
+import "../Serialization.sol";
+import "../Math.sol";
+import "../Standards/IERC223Token.sol";
+
+
+contract FeeDisbursal is
+    AccessControlled,
+    Reclaimable,
+    IFeeDisbursal,
+    IdentityRecord,
+    Serialization,
+    Math
+{
+
+    uint256 constant UINT256_MAX = 2**256 - 1;
+
+
+    ////////////////////////
+    // Types
+    ////////////////////////
+    struct Disbursal {
+        // snapshop ID of the pro-rata token, which will define which amounts to disburse against
+        uint256 snapshotId;
+        // amount of tokens to disburse
+        uint256 amount;
+        // address of the token used to determine the user pro rata amount, must be a snapshottoken, default is NEU
+        BasicSnapshotToken proRataToken;
+        // time after which claims to this token can be recycled
+        uint256 recycleAfterTimestamp;
+        // contract sending the disbursal
+        address disburser;
+    }
+
+    ////////////////////////
+    // Immutable state
+    ////////////////////////
+    Universe private UNIVERSE;
+    IIdentityRegistry private IDENTITY_REGISTRY;
+
+    ////////////////////////
+    // Mutable state
+    ////////////////////////
+
+    // map token addresses to a list of disbursal events of that token
+    mapping (address => Disbursal[]) private _disbursals;
+    // mapping to track what disbursals have already been paid out to which user
+    // token address => user address => disbursal progress
+    mapping (address => mapping(address => uint256)) _disbursalProgress;
+
+
+    ////////////////////////
+    // Constructor
+    ////////////////////////
+    constructor(Universe universe)
+        AccessControlled(universe.accessPolicy())
+        Reclaimable()
+        public
+    {
+        UNIVERSE = universe;
+        IDENTITY_REGISTRY = IIdentityRegistry(universe.identityRegistry());
+    }
+
+    ////////////////////////
+    // Public functions
+    ////////////////////////
+
+    /// @notice get the disbursal at a given index for a given token
+    /// @param token address of the claimable token
+    /// @param index until what index to claim to
+    function getDisbursal(address token, uint256 index)
```

Comment:
If you make iterator then also provide length somewhere. Or we iterate until we have 0 return struct?

Reply:
added a count function